#159987690 User should get verification email after signup

emmaadesile opened the pull request:
### What does this PR do?
- Implement the feature that sends a verification email to a user after successful signup.
- Confirms user's email after clicking on the verification link in the email.
- Resend email verification to user when they request for it.

### Description of Task to be completed?
- added two new endpoints to the routes
```
  POST /api/users/confirmation/:token
  POST /api/users/verify/resend-email
  ```
- user should get a verification email after successful signup
- user's email should be verified when they hit this route `/api/users/confirmation/:token`;
- user can also request that the verification email be resent when they make a post request to this route `/api/users/verify/resend-email`

### How should this be manually tested?
- After cloning the app
- create a `.env` file from `.envsample`
- run `npm install`
- Install and setup Postgres
- run `npm start:dev` to start the development server
- on postman, make a `POST` request to `/api/users` to signup a new user
- You should a message that a verification email has been sent to the user
- Copy the link from the email to make a `GET` request to `/api/users/confirmation/:token`
- You should get a message that the email has been confirmed
- Request that verification email be resent by making a `POST` request to  `/api/users/verify/resend-email`

### What are the relevant pivotal tracker stories?
[159987690](https://www.pivotaltracker.com/story/show/159987690)

lauragift21 committed:
feat(docs): Integrate Swagger for documenting API

- implement swagger-ui-express to enable documentation of API

[Delivers #159987686]

emekafredy committed:
feat(usersProfile): create endpoints for user profiles

- Create endpoints for users to view and edit profile
- Implement validations for user profile and login inputs
- Test endpoints for the users profile

[Delivers #159987699]

emmaadesile committed:
feat(verify-email): user should get verification email after signup

- user should get a verification email after signup
- user should be able to click the link to verify their email
- user's email should be confirmed after clicking the link
- emailverified field in Users Model should be changed to true after confirmation

[Delivers #159987690]

aknwosu committed:
Merge pull request #13 from andela/ft-document-api-swagger-159987686

#159987686 Implement API documentation with Swagger

aknwosu commented on server/controllers/emailVerificationController.js:
```diff
@@ -0,0 +1,79 @@
+            })
+          ))
+        }) 
+        .catch(error => console.log(error))
```
console.log? Properly catch errors sending appropriate message and status

aknwosu commented:
@emmaadesile Sync with @CEOehis to fix Travis error, standardize your object notations. e.g. use { expect } or {expect} consistently with { expect } being the preferred

emekafredy reviewed (commented):
Resolve issues with houndCI

aknwosu committed:
Merge pull request #18 from andela/ft-create-profile-159987699

#159987699 Users should be able to see other users' profile and update their own profile

emmaadesile committed:
ft(verify-email): merge remote commits from develop branch to local branch

emmaadesile committed:
ft(confirm-email): user should get verification email after signup

- user should a verification after successful signup
- user's email should be verified after clicking the link in the email
- verified field in the user's table should be changed to true

[Delivers 159987690]

emmaadesile committed:
Merge branch 'ft-verify-email-159987690' of https://github.com/andela/thor-ah into ft-verify-email-159987690

ft(confirm-meial): fix conflicts while pushing to remote branch

[Delivers 159987690]

emmaadesile committed:
ft(confirm-email): fix conflix in migrations file while pushing to remote branch

[Delivers 159987690]

emmaadesile committed:
ft(verify-email): fix conflicts while pushing to remote branch

emmaadesile committed:
ft(confirm-email): confirm user's email address

- write tests for endpoint that confirm's user's email

[Delivers 159987690]

emmaadesile committed:
ft(confirm-email): write tests for verify user's email  endpoint

[Delivers 159987690]

aknwosu commented on server/controllers/emailVerificationController.js:
```diff
@@ -0,0 +1,134 @@
+        if (!user) {
+          return res.status(404).json({
+            status: 'failed',
+            message: 'User does not exist in database'
```
message: 'User does not exist'

aknwosu commented on server/controllers/emailVerificationController.js:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        const emailToken = jwt.sign({ id: user.id }, secret, { expiresIn: '1h' });
+        const url = `http://localhost:3000/api/users/confirmation/${emailToken}`;
```
have this in your env file

aknwosu commented on server/controllers/emailVerificationController.js:
```diff
@@ -0,0 +1,134 @@
+        const msg = {
+          to: email,
+          from: 'noreply@authorshaven.com',
+          subject: 'Authors Haven Email Verification',
```
Authors' Haven Email Verification. apostrophe missing

aknwosu commented on server/controllers/users.js:
```diff
@@ -0,0 +1,216 @@
+import UserValidation from '../validation/users';
+import trimInput from '../utils/trim_input';
+
+const { User } = db;
```
import { User } from '../models'

aknwosu commented on server/controllers/users.js:
```diff
@@ -0,0 +1,216 @@
+          if (isMatch) {
+            return res.status(200).json({
+              message: 'Login successful',
+              user,
```
check the response object from others and make this correspond

aknwosu commented on server/controllers/users.js:
```diff
@@ -0,0 +1,216 @@
+              image: image || user.image,
+            }).then((updatedUser) => {
+              const { dataValues } = updatedUser;
+              return res.status(200).json({ dataValues });
```
correct all your response objects and standardize them

aknwosu replied:
really? anything as the response??

aknwosu commented on server/controllers/emailVerificationController.js:
```diff
@@ -0,0 +1,134 @@
+      const { id, email } = user;
+      const emailToken = jwt.sign({ email, id }, secret, { expiresIn: '1h' }); 
+
+      const url = `http://localhost:3000/api/users/confirmation/${emailToken}`;
```
if you hardcode the url here, what happens in production?

aknwosu commented:
I don't expect to still have to ask anyone to fix hound issues at this point
